symbolic-minidump/src/processor.rs

```diff
@@ -86,6 +87,290 @@ extern "C" {
+    /// Returns the position in the inner vector where the given range could be inserted, if that is possible.
+    fn free_slot(&self, range: &Range<A>) -> Option<usize> {
+        let index = match self.inner.binary_search_by_key(&range.end, |r| r.0.end) {
+            Ok(_) => return None,
+            Err(index) => index,
+        };
+
+        if index > 0 {
+            let before = &self.inner[index - 1];
+            if before.0.end > range.start {
+                return None;
+            }
+        }
+
+        match self.inner.get(index) {
+            Some(after) if after.0.start < range.end => None,
+            _ => Some(index),
+        }
+    }
+
+    /// Retrieves the range covering the given address and the associated contents.
+    pub fn get(&self, address: A) -> Option<&(Range<A>, E)> {
+        let entry = match self
+            .inner
+            .binary_search_by_key(&address, |range| range.0.end)
+        {
+            // This means inner(index).end == address => address might be covered by the next one
+            Ok(index) => self.inner.get(index + 1)?,
+            // This means that inner(index).end > address => this could be the one
+            Err(index) => self.inner.get(index)?,
+        };
+
+        (entry.0.start <= address).then(|| entry)
+    }
+
+    /// Retrieves the range covering the given address, allowing mutation.
+    pub fn get_mut(&mut self, address: A) -> Option<&mut (Range<A>, E)> {
+        let entry = match self
+            .inner
+            .binary_search_by_key(&address, |range| range.0.end)
+        {
+            // This means inner(index).end == address => address might be covered by the next one
+            Ok(index) => self.inner.get_mut(index + 1)?,
+            // This means that inner(index).end > address => this could be the one
+            Err(index) => self.inner.get_mut(index)?,
+        };
+
+        (entry.0.start <= address).then(|| entry)
+    }
+
+    /// Retrieves the contents associated with the given address.
+    pub fn get_contents(&self, address: A) -> Option<&E> {
+        self.get(address).map(|(_, contents)| contents)
+    }
+
+    /// Retrieves the contents associated with the given address, allowing mutation.
+    pub fn get_contents_mut(&mut self, address: A) -> Option<&mut E> {
+        self.get_mut(address).map(|(_, contents)| contents)
+    }
+
+    /// Returns true if the given address is covered by some range in the map.
+    pub fn contains(&self, address: A) -> bool {
+        self.get(address).is_some()
+    }
+}
+
+impl<A, E> Default for RangeMap<A, E> {
+    fn default() -> Self {
+        Self { inner: Vec::new() }
+    }
+}
+
+type NestedRangeMapEntry<A, E> = (Range<A>, E, Box<NestedRangeMap<A, E>>);
+
+/// A structure representing a tree of disjoint ranges with associated contents.
+#[derive(Debug)]
+pub struct NestedRangeMap<A, E> {
+    inner: Vec<NestedRangeMapEntry<A, E>>,
+}
+
+impl<A: Ord + Copy + fmt::Debug, E> NestedRangeMap<A, E> {
```

Comment:
I think its best practice in Rust land to avoid huge trait bounds that are not strictly needed. What I mean here specifically is:
* The `Debug` bound is only needed (and the only one needed) for the `from_vec_unchecked` method.
* The other methods need the `Ord + Copy` bounds.

No need to change this now, just to keep in mind. Especially the stdlib sometimes even puts the trait bounds on the methods themselves.

Reply:
I think I'm actually going to remove this method. It doesn't really serve much of a purpose.